test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/MemberExpressionCacheKeyComparerTest.cs

```diff
@@ -0,0 +1,169 @@
+
+        [Fact]
+        public void Equals_ReturnsTrue_ForDifferentInstances_OfSameExpression()
+        {
+            // Arrange
+            var key1 = GetKey(m => m.Value);
+            var key2 = GetKey(m => m.Value);
+
+            // Act & Assert
+            VerifyEquals(key1, key2);
+        }
+
+        [Fact]
+        public void Equals_ReturnsTrue_ForChainedMemberAccessExpressionsWithReferenceTypes()
+        {
+            // Arrange
+            var key1 = GetKey(m => m.TestModel2.Name);
+            var key2 = GetKey(m => m.TestModel2.Name);
+
+            // Act & Assert
+            VerifyEquals(key1, key2);
+        }
+
+        [Fact]
+        public void Equals_ReturnsTrue_ForChainedMemberAccessExpressionsWithNullableValueTypes()
+        {
+            // Arrange
+            var key1 = GetKey(m => m.NullableDateTime.Value.TimeOfDay);
+            var key2 = GetKey(m => m.NullableDateTime.Value.TimeOfDay);
+
+            // Act & Assert
+            VerifyEquals(key1, key2);
+        }
+
+        [Fact]
+        public void Equals_ReturnsTrue_ForChainedMemberAccessExpressionsWithValueTypes()
+        {
+            // Arrange
+            var key1 = GetKey(m => m.DateTime.Year);
+            var key2 = GetKey(m => m.DateTime.Year);
+
+            // Act & Assert
+            VerifyEquals(key1, key2);
+        }
+
+        [Fact]
+        public void Equals_ReturnsFalse_ForDifferentExpression()
+        {
+            // Arrange
+            var key1 = GetKey(m => m.Value);
+            var key2 = GetKey(m => m.TestModel2.Name);
+
+            // Act & Assert
+            VerifyNotEquals(key1, key2);
+        }
+
+        [Fact]
+        public void Equals_ReturnsFalse_WhenMemberIsAccessedThroughNullableProperty()
+        {
+            // Arrange
+            var key1 = GetKey(m => m.NullableDateTime.Value.Year);
+            var key2 = GetKey(m => m.DateTime.Year);
+
+            // Act
+            VerifyNotEquals(key1, key2);
+        }
+
+        [Fact]
+        public void Equals_ReturnsFalse_WhenMemberIsAccessedThroughDifferentModels()
+        {
+            // Arrange
+            var key1 = GetKey<TestModel2, int>(m => m.Id);
+            var key2 = GetKey(m => m.TestModel2.Id);
+
+            // Act
+            VerifyNotEquals(key1, key2);
+        }
+
+        [Fact]
+        public void Equals_ReturnsFalse_WhenMemberIsAccessedThroughConstantExpression()
+        {
+            // Arrange
+            var testModel = new TestModel2 { Id = 1 };
+            var key1 = GetKey(m => testModel.Id);
+            var key2 = GetKey<TestModel2, int>(m => m.Id);
+
+            // Act
+            VerifyNotEquals(key1, key2);
+        }
+
+        private void VerifyEquals(MemberExpressionCacheKey key1, MemberExpressionCacheKey key2)
+        {
+            var hashCode1 = Comparer.GetHashCode(key1);
+            var hashCode2 = Comparer.GetHashCode(key2);
+
+            var equals = Comparer.Equals(key1, key2);
```

Comment:
Make this `Assert.Equal(key1, key2, Comparer)` and do it before getting the hash codes